File: Stack/Java/Sum of Subarray Ranges.java
// Problem link: https://leetcode.com/problems/sum-of-subarray-ranges/description/
// Time: O(n)
// Space: O(n)
class Solution {
    public long subArrayRanges(int[] nums) {
        return this.sumSub(nums, true) - this.sumSub(nums, false);
    }

    private long sumSub(int[] nums, boolean flag) {
        Stack<int[]> stack = new Stack<>();
        long result = 0;
        int left, right;

        for(int idx = 0; idx < nums.length; idx++) {
            while(!stack.isEmpty() && (flag ? stack.peek()[1] < nums[idx] : stack.peek()[1] > nums[idx])) {
                int[] popPair = stack.pop();
                left = stack.isEmpty() ? popPair[0] + 1 : popPair[0] - stack.peek()[0];
                right = idx - popPair[0];
                result += (long)right * left * popPair[1];  
            }
            stack.push(new int[] {idx, nums[idx]});
        }

        while(!stack.isEmpty()) {
            int[] popPair = stack.pop();
            left = stack.isEmpty() ? popPair[0] + 1 : popPair[0] - stack.peek()[0];
            right = nums.length - popPair[0];
            result += (long)right * left * popPair[1];  
        }

        return result;
    }
}

File: Stack/Python/Sum of Subarray Ranges.py
# Problem link: https://leetcode.com/problems/sum-of-subarray-ranges/description/
# Time: O(n)
# Space: O(n)
class Solution:
    def subArrayRanges(self, nums: List[int]) -> int:
        return self.sumSub(nums, 0) - self.sumSub(nums, 1)
    
    def sumSub(self, arr, flag):
        result, stack = 0, []

        for idx, val in enumerate(arr):
            while stack and (stack[-1][1] > val if flag else stack[-1][1] < val):
                index, num = stack.pop()
                left = index - stack[-1][0] if stack else index + 1
                right = idx - index
                result +=left * right * num
            stack.append((idx, val))
        
        while stack:
            index, num = stack.pop()
            left = index - stack[-1][0] if stack else index + 1
            right = len(arr) - index
            result += left * right * num
        
        return result
